### DISCUSSION TITLE
Offical Dockerfile fails when copying /dist folder, says it doesn't exist.

### DISCUSSION THREAD

**Discussion by MeikyuuTrader** [directus/directus#15199] [Archive]
Trying to build the [official dockerfile](https://github.com/directus/directus/blob/main/docker/Dockerfile) but I keep getting an error on this line `COPY /dist .` with the error

```
Step 5/22 : COPY /dist .
COPY failed: file not found in build context or excluded by .dockerignore: stat dist: file does not exist
```

My dockerfile is copy and paste of the linked official dockerfile, fails on local and using render deployment when building the dockerfile. 

It's currently using Node 16-alpine, is the issue that node-16-alpine doesn't have a /dist folder, thus this error occurs? Tried `RUN find /dist` and it says that doesn't exist.

My current home folder only contains the dockerfile itself.

# Edit
Mainly an issue with no understanding how the dockerfile is supposed to work. Assumed just copying and pasting the dockerfile would have everything run, see answer for real instructions. 

# Packages and Versions
Docker version 20.10.14, build a224086349

**Comment by azrikahar** [comment]
> My current home folder only contains the dockerfile itself.

The Dockerfile will require additional steps and build artifacts before that, so this will not work.

You can refer to the workflow that is being used here: https://github.com/directus/directus/blob/main/.github/workflows/release.yml

A summary should look something like:

1. Clone the repo

2. Run `pnpm install` 

3. Run `pnpm -r build` (Highly recommend referring step 1 to step 3 via the official docs here: https://docs.directus.io/contributing/running-locally.html)

4. Run `node docker/pack` (This should be what is responsible to create `/dist` in the project root)

   https://github.com/directus/directus/blob/ef939072cc0b41575f9f0507e5798ea2d645640b/docker/pack.js#L24

---

That being said, perhaps you can share what are you trying to do with the Dockerfile? Is there any particular reasons to not use the [official docker image](https://hub.docker.com/r/directus/directus)? If the goal is to _extend_ the image, rather than modifying something inherent in the official Dockerfile, you can also try creating a new Dockerfile that is based on the official image, such as:

```Dockerfile
FROM directus/directus:9.16.1

# your additional Dockerfile steps here
```

**Reply by MeikyuuTrader** [reply]
I can't seem to connect to directus when using the image from a dockerfile.

```
FROM directus/directus:9.16.1

ENV \
  KEY="KEY" \
  SECRET="SECRET"

EXPOSE 8055
```

After running, it'd eventually show `Server started at http://0.0.0.0:8055` but that doesn't connect.

**Reply by azrikahar** [reply]
Can you share how you are running this built image? Did you bind the port? Did you refer to the "Installing with Docker" section of the official Directus docs? Highly recommend checking it out: https://docs.directus.io/self-hosted/installation/docker.html

Example docker run command from the docs:

```shell
# Make sure to change sensitive values (KEY, SECRET, ...) in production
docker run \
  -p 8055:8055 \
  -e KEY=255d861b-5ea1-5996-9aa3-922530ec40b1 \
  -e SECRET=6116487b-cda1-52c2-b5b5-c8022c45e263 \
  directus/directus
```

Do note that `EXPOSE 8055` doesn't really expose said port, it's just more of a documentation type of deal. Quote from the official Docker docs:

> The `EXPOSE` instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published. To actually publish the port when running the container, use the `-p` flag on docker run to publish and map one or more ports, or the `-P` flag to publish all exposed ports and map them to high-order ports.

Link for above quote: https://docs.docker.com/engine/reference/builder/#expose

**Reply by MeikyuuTrader** [reply]
Ah, that's why then. My bad, I didn't realise expose didn't expose ports properly. Everything works fine then, thanks for the help.
